### PR TITLE
Sync done threads when polling for upload status

### PR DESCRIPTION
**Pull request by omerzi**
- [x] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-cli-core/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----
- Using the new Set implementation from:
https://github.com/jfrog/gofrog/pull/29

- Fixing an issue when the source sends 'DONE' while there is a connectivity issue between the source and the CLI. The 'DONE' get lost; therefore, the progress was stuck waiting for the 'DONE' for these chunks.

- Some changes in the PR are not related to the transfer files, They are included in order to pass the 'Static-Analysis' for Go v1.18 (e.g. deprecated function strings.Title())

### CODE REVIEW

**Comment by yahavi** [comment]
Pending on https://github.com/jfrog/data-transfer/pull/33

File: artifactory/commands/transferfiles/srcpluginservice.go [hunk 1]
```diff
@@ -34,19 +35,22 @@ func (sup *srcUserPluginService) IsDryRun() bool {
 	return false
 }
 
-func (sup *srcUserPluginService) getUploadChunksStatus(ucStatus UploadChunksStatusBody) (UploadChunksStatusResponse, error) {
+func (sup *srcUserPluginService) syncChunks(ucStatus *UploadChunksStatusBody) (UploadChunksStatusResponse, error) {
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/srcpluginservice.go:38]: * Let's send the `ucStatus` argument by value and not by reference, to reduce the concern that this function will modify it.
* Let's remive the following clean up lines, and let the caller function handle the cleanup - 
```
	ucStatus.AwaitingStatusChunks = nil
	ucStatus.ChunksToDelete = nil
```
* The cleanup on the caller function ca be done by initializing a new `UploadChunksStatusBody` struct abd only if there no errors returned from this function.

File: artifactory/commands/utils/configfile.go [hunk 2]
```diff
@@ -298,7 +300,7 @@ func getIncludeExcludePatterns(patternType string) string {
 	var patterns []string
 	patternNum := 1
 	for {
-		newPattern := AskString("", strings.Title(patternType)+" pattern "+strconv.Itoa(patternNum)+" (leave empty to continue):", true, false)
+		newPattern := AskString("", cases.Title(language.Und, cases.NoLower).String(patternType)+" pattern "+strconv.Itoa(patternNum)+" (leave empty to continue):", true, false)
```
Comment (eyalbe4) [inline artifactory/commands/utils/configfile.go:303]: What is the reason for this change?

File: general/cisetup/jenkinsfiledslgenerator.go [hunk 3]
```diff
@@ -190,7 +192,7 @@ func generateRtConfigSteps(techInfo *TechnologyInfo, rtUrl string) string {
 	buildType := string(techInfo.Type)
 	resolverId := fmt.Sprintf(resolverIdTemplate, strings.ToUpper(buildType))
 	deployerId := fmt.Sprintf(deployerIdTemplate, strings.ToUpper(buildType))
-	return fmt.Sprintf(rtConfigServerStepTemplate, ConfigServerId, rtUrl, strings.Title(buildType), deployerId, deployerRepos, resolverId, resolverRepos)
+	return fmt.Sprintf(rtConfigServerStepTemplate, ConfigServerId, rtUrl, cases.Title(language.Und, cases.NoLower).String(buildType), deployerId, deployerRepos, resolverId, resolverRepos)
```
Comment (eyalbe4) [inline general/cisetup/jenkinsfiledslgenerator.go:195]: What is the reaon for this change?

File: artifactory/commands/transferfiles/utils.go [hunk 4]
```diff
@@ -151,7 +153,12 @@ func pollUploads(phaseBase *phaseBase, srcUpService *srcUserPluginService, uploa
 						continue
 					}
 				}
-				curTokensBatch.UuidTokens = removeTokenFromBatch(curTokensBatch.UuidTokens, chunk.UuidToken)
+				err := awaitingStatusChunksSet.Remove(chunk.UuidToken)
+				if err != nil {
+					log.Error("removal from upload status set issue occurred: " + err.Error())
+					continue
+				}
+				curTokensBatch.ChunksToDelete = append(curTokensBatch.ChunksToDelete, chunk.UuidToken)
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:161]: Let's add a comment here explaining why we're doing this. Let's make it clear for the random reader what `ChunksToDelete` actually means and why it is needed.

File: artifactory/commands/transferfiles/utils.go [hunk 5]
```diff
@@ -118,17 +120,17 @@ func pollUploads(phaseBase *phaseBase, srcUpService *srcUserPluginService, uploa
 		if progressbar != nil {
 			progressbar.SetRunningThreads(curProcessedUploadChunks)
 		}
-		curTokensBatch.fillTokensBatch(uploadTokensChan)
-
-		if len(curTokensBatch.UuidTokens) == 0 {
+		fillTokensBatch(awaitingStatusChunksSet, uploadTokensChan)
+		if awaitingStatusChunksSet.Size() == 0 && len(curTokensBatch.ChunksToDelete) == 0 {
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:124]: Let's explain here to the random reader what `awaitingStatusChunksSet` stores and also what `ChunksToDelete`, and then what does it mean when the size for both is zero.

File: artifactory/commands/transferfiles/utils.go [hunk 6]
```diff
@@ -118,17 +120,17 @@ func pollUploads(phaseBase *phaseBase, srcUpService *srcUserPluginService, uploa
 		if progressbar != nil {
 			progressbar.SetRunningThreads(curProcessedUploadChunks)
 		}
-		curTokensBatch.fillTokensBatch(uploadTokensChan)
-
-		if len(curTokensBatch.UuidTokens) == 0 {
+		fillTokensBatch(awaitingStatusChunksSet, uploadTokensChan)
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:123]: Let's explain what `fillTokensBatch` does.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: artifactory/commands/transferfiles/srcpluginservice.go [hunk 7]
```diff
@@ -34,19 +35,22 @@ func (sup *srcUserPluginService) IsDryRun() bool {
...
 
 	httpDetails := sup.GetArtifactoryDetails().CreateHttpClientDetails()
 	utils.SetContentType("application/json", &httpDetails.Headers)
-	resp, body, err := sup.client.SendPost(sup.GetArtifactoryDetails().GetUrl()+pluginsExecuteRestApi+"getUploadChunksStatus", content, &httpDetails)
+	resp, body, err := sup.client.SendPost(sup.GetArtifactoryDetails().GetUrl()+pluginsExecuteRestApi+syncChunks, content, &httpDetails)
 	if err != nil {
 		return UploadChunksStatusResponse{}, err
 	}
 
+	// clear arrays for the next request body
```
[the remark this answers is not in the document]
Comment (omerzi) [reply artifactory/commands/transferfiles/srcpluginservice.go:51]: done

File: artifactory/commands/transferfiles/srcpluginservice.go (hunk 1 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/srcpluginservice.go:38]: done

File: general/cisetup/jenkinsfiledslgenerator.go (hunk 3 above)
Comment (omerzi) [reply general/cisetup/jenkinsfiledslgenerator.go:195]: We updated CLI-Core to Go v1.18, this caused Static Analysis failure, because 'strings.Title()' is now deprecated in Go v1.18, and 'cases' replace it:
https://github.com/jfrog/jfrog-cli-core/actions/runs/2942074856

File: artifactory/commands/utils/configfile.go (hunk 2 above)
Comment (omerzi) [reply artifactory/commands/utils/configfile.go:303]: We updated CLI-Core to Go v1.18, this caused Static Analysis failure, because 'strings.Title()' is now deprecated in Go v1.18, and 'cases' replace it:
https://github.com/jfrog/jfrog-cli-core/actions/runs/2942074856

File: artifactory/commands/transferfiles/utils.go (hunk 4 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/utils.go:161]: done

File: artifactory/commands/transferfiles/utils.go (hunk 6 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/utils.go:123]: done

File: artifactory/commands/transferfiles/utils.go (hunk 5 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/utils.go:124]: done

File: general/cisetup/jenkinsfiledslgenerator.go (hunk 3 above)
Comment (eyalbe4) [reply general/cisetup/jenkinsfiledslgenerator.go:195]: Got it. NP!

File: artifactory/commands/transferfiles/transfer.go [hunk 8]
```diff
@@ -198,6 +198,13 @@ func (tdc *TransferFilesCommand) transferSingleRepo(sourceRepoKey string, target
 			return
 		}
 	}
+	// Performing a reset on the source if chunks were not removed
+	runningNodes, err := getRunningNodes(tdc.sourceServerDetails)
+	if err != nil {
+		log.Error(err)
+	} else {
+		stopTransferOnArtifactoryNodes(srcUpService, runningNodes)
+	}
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/transfer.go:207]: Notice the above code is duplicated twice in the code.
We should theregore move into a util function named `stopTransferInArtifactory` which recieves no arguments. This utility function should return an error, which is checked here. amd also in the other functions that will use this utility function.

Also, the existing `stopTransferOnArtifactoryNodes` should be renamed to `stopTransferInArtifactoryNodes`.

I also suggest running this lceanup before initiating a new phase (any new phase) and not after finishing all phases for a repository.

File: artifactory/commands/transferfiles/transfer.go [hunk 9]
```diff
@@ -198,6 +198,13 @@ func (tdc *TransferFilesCommand) transferSingleRepo(sourceRepoKey string, target
 			return
 		}
 	}
+	// Performing a reset on the source if chunks were not removed
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/transfer.go:201]: // Performing a reset on the source if chunks were not removed
-->
// Ensure the data structure which stores the upload tasks on Artifactory's side is wiped clean, in case some of the requests to delete handles tasks sent by JFrog CLI did not reach Artifactory.

File: artifactory/commands/transferfiles/utils.go [hunk 10]
```diff
@@ -119,21 +121,29 @@ func pollUploads(phaseBase *phaseBase, srcUpService *srcUserPluginService, uploa
 		if progressbar != nil {
 			progressbar.SetRunningThreads(curProcessedUploadChunks)
 		}
-		curTokensBatch.fillTokensBatch(uploadTokensChan)
-
-		if len(curTokensBatch.UuidTokens) == 0 {
+		// Each uploading thread receive a token from the source via the uploadTokensChan, so this go routine can poll on it's status
+		fillTokensBatch(awaitingStatusChunksSet, uploadTokensChan)
+		// awaitingStatusChunksSet is used to keep all the uploaded chunks tokens in order to request their upload status from the source
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:126]: * Let's add a new-line before the beginning of this description.
* Missing period at the end of the sentence.

File: artifactory/commands/transferfiles/utils.go [hunk 11]
```diff
@@ -119,21 +121,29 @@ func pollUploads(phaseBase *phaseBase, srcUpService *srcUserPluginService, uploa
...
+		// ChunksToDelete is used to notify the source that these chunks can be deleted from the source's status map
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:127]: Missing period at the end of the sentence.

File: artifactory/commands/transferfiles/utils.go [hunk 12]
```diff
@@ -119,21 +121,29 @@ func pollUploads(phaseBase *phaseBase, srcUpService *srcUserPluginService, uploa
...
+		// After we receive 'DONE', we inform the source that the 'DONE' message has been received, and it no longer has to keep those chunk uuids.
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:128]: ...chunk uuids. --> ...chunks' UUIDs.

File: artifactory/commands/transferfiles/utils.go [hunk 13]
```diff
@@ -119,21 +121,29 @@ func pollUploads(phaseBase *phaseBase, srcUpService *srcUserPluginService, uploa
...
+		// When both ChunksToDelete and awaitingStatusChunksSet length is zero means that all the tokens has been uploaded,
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:129]: ...is zero means ... --> ...is zero, it means ...

Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:129]: ...has been uploaded, --> ...are uploaded,

File: artifactory/commands/transferfiles/utils.go [hunk 14]
```diff
@@ -119,21 +121,29 @@ func pollUploads(phaseBase *phaseBase, srcUpService *srcUserPluginService, uploa
...
+		// we received 'DONE' for all of them, and we notified the source that they can be deleted from the memory
```
Comment (eyalbe4) [inline artifactory/commands/transferfiles/utils.go:130]: Missing period at the end of the sentence.

File: go.mod [hunk 15]
```diff
@@ -28,6 +28,8 @@ require (
 	gopkg.in/yaml.v2 v2.4.0
 )
 
+require golang.org/x/text v0.3.7
```
Comment (eyalbe4) [inline go.mod:31]: Can we add this dependency declaration to the above `require` section?

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: artifactory/commands/transferfiles/transfer.go (hunk 8 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/transfer.go:207]: Done, but the function must get the serverDetails and the srcUpService as arguments

File: artifactory/commands/transferfiles/transfer.go (hunk 9 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/transfer.go:201]: Done

File: artifactory/commands/transferfiles/utils.go (hunk 10 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/utils.go:126]: done

File: artifactory/commands/transferfiles/utils.go (hunk 11 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/utils.go:127]: done

File: artifactory/commands/transferfiles/utils.go (hunk 12 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/utils.go:128]: done

File: go.mod (hunk 15 above)
Comment (omerzi) [reply go.mod:31]: done

File: artifactory/commands/transferfiles/utils.go (hunk 14 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/utils.go:130]: done

File: artifactory/commands/transferfiles/utils.go (hunk 13 above)
Comment (omerzi) [reply artifactory/commands/transferfiles/utils.go:129]: done

Comment (omerzi) [reply artifactory/commands/transferfiles/utils.go:129]: done

File: artifactory/commands/transferfiles/transfer.go [hunk 16]
```diff
@@ -192,7 +194,12 @@ func (tdc *TransferFilesCommand) transferSingleRepo(sourceRepoKey string, target
 		if tdc.ShouldStop() {
 			return
 		}
-
+		// Ensure the data structure which stores the upload tasks on Artifactory's side is wiped clean,
+		// in case some of the requests to delete handles tasks sent by JFrog CLI did not reach Artifactory.
+		err = stopTransferInArtifactory(tdc.sourceServerDetails, srcUpService)
```
Comment (yahavi) [inline artifactory/commands/transferfiles/transfer.go:199]: I think it would be better to put in a defer before the "for" loop. Something like:
```go
defer func ()  {
	if tdc.ShouldStop() {
		// stopTransferInArtifactory already invoked
		return
	}
	stopTransferInArtifactory(tdc.sourceServerDetails, srcUpService)
	if err != nil {
		log.Error(err)
	}
}
```

**Review by yahavi — CHANGES_REQUESTED** [review]
Let's also fix the tests
